components/mediation/data-publishers/org.wso2.micro.integrator.observability/src/main/java/org/wso2/micro/integrator/obsrvability/handler/metrics/publisher/MetricHandler.java

```diff
@@ -0,0 +1,480 @@
+    private void incrementInboundEndPointCount(String name) {
+        Map<String, String[]> metricCountMap = new HashMap();
+        metricCountMap.put(MetricConstants.INBOUND_ENDPOINT_REQUEST_COUNT_TOTAL, new String[]{name,
+                MetricConstants.INBOUND_ENDPOINT});
+        metricReporterInstance.incrementCount(MetricConstants.INBOUND_ENDPOINT_REQUEST_COUNT_TOTAL, metricCountMap);
+    }
+
+    /**
+     * Increment the error request count received by a proxy service.
+     *
+     * @param name The metric name
+     */
+    private void incrementProxyErrorCount(String name) {
+        Map<String, String[]> metricCountMap = new HashMap();
+        metricCountMap.put(MetricConstants.PROXY_REQUEST_COUNT_ERROR_TOTAL, new String[]{name,
+                SynapseConstants.PROXY_SERVICE_TYPE});
+        metricReporterInstance.incrementCount(MetricConstants.PROXY_REQUEST_COUNT_ERROR_TOTAL, metricCountMap);
+    }
+
+    /**
+     * Increment the error request count received by an api.
+     *
+     * @param name The metric name
+     */
+    private void incrementApiErrorCount(String name, String apiInvocationUrl) {
+        Map<String, String[]> metricCountMap = new HashMap();
+        metricCountMap.put(MetricConstants.API_REQUEST_COUNT_ERROR_TOTAL, new String[]{name,
+                SynapseConstants.FAIL_SAFE_MODE_API, apiInvocationUrl});
+        metricReporterInstance.incrementCount(MetricConstants.API_REQUEST_COUNT_ERROR_TOTAL, metricCountMap);
+    }
+
+    /**
+     * Increment the error request count received by an inbound endpoint.
+     *
+     * @param name The metric name
+     */
+    private void incrementInboundEndpointErrorCount(String name) {
+        Map<String, String[]> metricCountMap = new HashMap();
+        metricCountMap.put(MetricConstants.INBOUND_ENDPOINT_REQUEST_COUNT_ERROR_TOTAL, new String[]{name,
+                MetricConstants.INBOUND_ENDPOINT});
+        metricReporterInstance.incrementCount(MetricConstants.INBOUND_ENDPOINT_REQUEST_COUNT_ERROR_TOTAL,
+                metricCountMap);
+    }
+
+    /**
+     * Get the api name.
+     *
+     * @param contextPath The api context path
+     * @param synCtx The Synapse Message Context
+     * @return String The api name
+     */
+    private String getApiName(String contextPath, MessageContext synCtx) {
+        String apiName = null;
+        for (API api : synCtx.getEnvironment().getSynapseConfiguration().getAPIs()) {
+            if (restUtils.getRESTApiName(contextPath, api.getContext())) {
+                apiName = api.getName();
+                if (api.getVersionStrategy().getVersion() != null && !"".equals(api.getVersionStrategy().
+                        getVersion())) {
+                    apiName = apiName + ":v" + api.getVersionStrategy().getVersion();
+                }
+                synCtx.setProperty(MetricConstants.IS_ALREADY_PROCESSED_REST_API, true);
+                synCtx.setProperty(MetricConstants.PROCESSED_API, api);
+            }
```

Comment:
why are we setting properties in getApiName method. let's rename accordingly. and I feel this should go to synapse and not here.

Reply:
We are setting these properties to avoid duplicate name generation in the RESTRequestHandler. We set these properties here and check whether the APIName is already identified and avoid the duplication in name creation.